python/taichi/lang/error.py

```diff
@@ -0,0 +1,59 @@
+import functools
+import traceback
+from colorama import Fore, Style
+import sys
+
+
+def no_traceback(foo):
+    '''
+    Don't include the decorated frame in traceback, if ti.set_excepthook() was called.
+    '''
+    @functools.wraps(foo)
+    def wrapped(*args, **kwargs):
+        _taichi_skip_traceback = 2
+        return foo(*args, **kwargs)
+
+    return wrapped
+
+
+def enable_excepthook():
+    def excepthook(exctype, value, tb):
+        skip = 0
+        back = 4
+        forward = 2
+        bar = f'{Fore.LIGHTBLACK_EX}{"-"*44}{Fore.RESET}'
+        print(
+            f'{Fore.LIGHTBLACK_EX}========== Taichi Stack Traceback =========={Fore.RESET}'
+        )
+        for frame, lineno in traceback.walk_tb(tb):
+            name = frame.f_code.co_name
+            filename = frame.f_code.co_filename
+            if '_taichi_skip_traceback' in frame.f_locals:
+                skip = frame.f_locals['_taichi_skip_traceback']
+            if skip > 0:
+                skip -= 1
+                continue
+            print(
+                f'In {Fore.LIGHTYELLOW_EX}{name}{Fore.RESET}() at {Fore.LIGHTMAGENTA_EX}{filename}{Fore.RESET}:{Fore.LIGHTCYAN_EX}{lineno}{Fore.RESET}:\n{bar}'
+            )
+            with open(filename) as f:
```

Comment:
Opt (maybe another PR, this one is not trivial already..): maybe we can cache the file content?

Reply:
I think the ultimate solution is `import inspect`..